storage_thumbnail: do not store url

simahawk opened the pull request:
The URL for all storage files is normally computed via backend.
This ensures that if you have different conf by env
you won't mess up w/ images coming from other envs (eg: prod).

This change makes sure that the URL which is used for the thumbs
is always computed by the backend.

simahawk committed:
storage_thumb: do not store url

The URL for all storage files is normally computed via backend.
This ensures that if you have different conf by env
you won't mess up w/ images coming from other envs (eg: prod).

This change makes sure that the URL which is used for the thumbs
is always computed by the backend.

simahawk committed:
storage_thumbnail: use FakeModelLoader for tests

simahawk committed:
storage_image tests: use SavepointComponentCase

simahawk commented:
@sebastienbeau @lmignon @bguillot what was the reason for storing images' url?
Can you check my `TODO`s in this PR?

simahawk commented on storage_image/models/storage_image.py:
```diff
@@ -47,6 +47,7 @@ def create(self, vals):
             vals["backend_id"] = self._get_default_backend_id()
         # When using the widget image_url, the create will pass the data
         # in the "url" field. We map it to the data field
+        # TODO: WHY??? I don't see any handling of `data` as url in storage.file
```
Would be nice to know why this was needed :)

simahawk committed:
Add odoo_test_helper to test deps

simahawk commented:
@sebastienbeau @lmignon @bguillot or @rousseldenis anyone? :pray:

lmignon commented on storage_image_product/models/product_product.py:
```diff
@@ -11,6 +11,7 @@ class ProductProduct(models.Model):
     # small and medium image are here to replace
     # native image field on form and kanban
     variant_image_small_url = fields.Char(
+        # TODO: why store?
```
store=True is probably required to avoid perf issues :thinking:

simahawk replied:
@lmignon yeah but at this stage the thumbs are already there and the computation should not take that much.
Maybe, it has a big impact on the kanban view?
BUT storing the URL is really a pain when you restore databases across different envs.

Ahhh, while thinking again on this (right now) I think we should store the relation to the thumbnails instead of the URL.
This way the computation should should be much faster. What do you think?

lmignon reviewed (approved):
LGTM (code review only)

simahawk committed:
storage_image_product: do not store images URLs

Storing URLs is wrong because they depend on the backend.
If backend rules change or environment changes
URLs will be outdated.
Instead, store the main image out of which we can compute the URLs.

simahawk committed:
storage_image_product: do not import tests common

simahawk commented:
@lmignon could you pass by again? I've refactored product image url computation in the commit `storage_image_product: do not store images URLs`.

CC @JordiBForgeFlow @HviorForgeFlow

simahawk committed:
storage_image_product: fix warning on duplicated labels

MiquelRForgeFlow commented:
@simahawk could you squash some commits? AFAIK, it should only be 2 commits: the refactor of tests and the non storing urls... This will give a clear understanding of the overall changes.

lmignon commented on storage_image_product/models/product_product.py:
```diff
@@ -8,24 +10,22 @@
+    )
+    # small and medium image are here to replace
+    # native image field on form and kanban
+    variant_image_small_url = fields.Char(related="main_image_id.image_small_url")
```
@simahawk IMO it's important to keep the image urls stored on the product template and on the variant to avoid extra queries when the product kanban view is displayed. 
If you change your env, you are not sure at all that your images exist into the other env. In this case, the new computed url could not exist.

simahawk replied:
@lmignon but THIS is the problem. If you change environment you should not see the images IF they are not there. Otherwise, what happens is that you expect to have those images in the system but you don't. Then the customer or a functional colleague comes to you and asks "hey, I see my images in Odoo, why I don't see them in the shop (or any other service)?".

simahawk replied:
also. if you need to change CDN URL on production, it won't have any effect on those URLs. That sucks IMO. One of the key point about storage files is that their URL is controlled by the backend.

lmignon replied:
> @lmignon but THIS is the problem. If you change environment you should not see the images IF they are not there. Otherwise, what happens is that you expect to have those images in the system but you don't. Then the customer or a functional colleague comes to you and asks "hey, I see my images in Odoo, why I don't see them in the shop (or any other service)?".

If you change environment and you don't want to see the image, you should apply a script to remove   `product.image.relation` and related `storage.image`and to set urls to blank.
Sometime it's nice to be able to see the images from an other env. (ex test env can see image from prod env but is not allowed to modify these items.)

lmignon replied:
> also. if you need to change CDN URL on production, it won't have any effect on those URLs. That sucks IMO. One of the key point about storage files is that their URL is controlled by the backend.

I prefer to write a script to upgrade urls by hand than fighting against performance issue... :smirk:

We could also add a listener on the write of `storage.backend` to update image urls if the URL changes on the backend in a efficient way (SQL ...)

simahawk replied:
> Sometime it's nice to be able to see the images from an other env. (ex test env can see image from prod env but is not allowed to modify these items.)

That's another kind of problem. We can work on it. Also, setting the production CDN on other envs won't have any impact on what you can write -> easy to do, fast result, no script / black magic needed.

> I prefer to write a script to upgrade urls by hand than fighting against performance issue... smirk

I think you are not considering the changes here: very likely perf were bad because of full awakening on several m2m fields: 

* rel w/ images
* rel w/ attribute values from the image relation
* rel w/ product attribute values on variants

a lot of m2m loaded every single time which we know to be very slow!
With this change you have only a m2o to look for.

Furthermore, the URL before my refactoring on https://github.com/OCA/storage/pull/88 was fully computed. The most expensive thing being the slug, which is now stored.

And now that I think about this again I think I'm going to store the main thumbs relation as well because to compute the thumb URL we always loop on all the thumbs and it can be expensive.

simahawk replied:
@lmignon so, in [this commit](https://github.com/OCA/storage/pull/94/commits/596dcf7a73fd4da9fd4798c4ef964a0f1350e4af)  `fixup! storage_thumb: do not store url` I've made the main thumbs computed. This ensures that the URL computation is very fast. I've tried on a kanban w/ thousands of records.

I've done a test by playing around w/ storage backend server env params and - as expected - if you keep the CDN from production (or whatever env the db is coming from) you'll see the images w/out any problem.

The only issue that I've found is that `base_url_for_files` is not refreshed when changing server env params as the change is not detected. Hence, I added a btn to refresh the base URL on the fly (this affects only the backend, no storage file computation is involved).

Finally, I made the 2 thumbs fields visible and editable so if by chance you want to take control of which image is used as thumbs, you can do it. No additional cost here.

Sounds good?

simahawk commented:
> @simahawk could you squash some commits? AFAIK, it should only be 2 commits: the refactor of tests and the non storing urls... This will give a clear understanding of the overall changes.

hmm that's the contrary in theory: all my commits are atomic and you can see from each commit what I changed exactly.

simahawk committed:
storage_file: allow base URL refresh via UI

simahawk committed:
storage_file: make url compute more robust

If any piece of the URL has a trailing slash, the URL might be wrong.

lmignon reviewed (approved):
Thank you for all your hard work @simahawk  You enhancements should improve the UX without introducing side effects. 
LGTM (Code review only) 
Unfortunately travis is :red_circle:

simahawk commented:
build is :green_circle: 

@MiquelRForgeFlow mind to give a review?

MiquelRForgeFlow commented:
sorry, I won't review. hope @HviorForgeFlow has some time :)

simahawk commented:
/ocabot merge minor
